File: aoc2025/go/challenges/day9.go
package challenges

import (
	"math"
	"slices"
	"strconv"
	"strings"

	"aoc/grid"
	"aoc/utilities"

	logger "github.com/Bparsons0904/goLogger"
)

type Interval struct {
	Start, End int
}

func Day9() {
	log := logger.New("Day9")

	filename := "day9.part1"
	redTiles := getRedTiles(filename)

	timer := log.Timer("Part 1 Timer")
	largestAreaPart1 := getLargestArea(redTiles)
	timer()

	timer = log.Timer("Part 2 Timer getInsideIntervals")
	intervals := getInsideIntervals(redTiles)
	timer()

	timer = log.Timer("Part 2 Original Timer")
	largestAreaPart2 := getLargestAreaWithIntervals(redTiles, intervals)
	timer()

	timer = log.Timer("Part 2 Optimized Timer")
	largestAreaPart2Optimized := getLargestAreaWithIntervalsOptimized(redTiles, intervals)
	timer()

	log.Info(
		"Results",
		"Part 1",
		largestAreaPart1,
		"Part 2",
		largestAreaPart2,
		"Part 2 Optimized",
		largestAreaPart2Optimized,
	)
}

func getInsideIntervals(redTiles []grid.Point) map[int][]Interval {
	log := logger.New("Day9").With("Function", "getInsideIntervals")

	type VerticalEdge struct {
		X          int
		MinY, MaxY int
	}
	type HorizontalEdge struct {
		Y          int
		MinX, MaxX int
	}

	var verticalEdges []VerticalEdge
	var horizontalEdges []HorizontalEdge

	timer := log.Timer("build edges")
	for i := range redTiles {
		current := redTiles[i]
		next := redTiles[(i+1)%len(redTiles)]

		if current.X == next.X {
			minY, maxY := current.Y, next.Y
			if minY > maxY {
				minY, maxY = maxY, minY
			}
			verticalEdges = append(
				verticalEdges,
				VerticalEdge{X: current.X, MinY: minY, MaxY: maxY},
			)
		} else {
			minX, maxX := current.X, next.X
			if minX > maxX {
				minX, maxX = maxX, minX
			}
			horizontalEdges = append(horizontalEdges, HorizontalEdge{Y: current.Y, MinX: minX, MaxX: maxX})
		}
	}
	timer()

	minY, maxY := redTiles[0].Y, redTiles[0].Y
	for _, p := range redTiles {
		if p.Y < minY {
			minY = p.Y
		}
		if p.Y > maxY {
			maxY = p.Y
		}
	}

	timer = log.Timer("build intervals")
	intervals := make(map[int][]Interval)

	for y := minY; y <= maxY; y++ {
		var crossings []int

		for _, edge := range verticalEdges {
			if y >= edge.MinY && y < edge.MaxY {
				crossings = append(crossings, edge.X)
			}
		}

		slices.Sort(crossings)

		var rowIntervals []Interval
		for i := 0; i+1 < len(crossings); i += 2 {
			rowIntervals = append(rowIntervals, Interval{Start: crossings[i], End: crossings[i+1]})
		}

		for _, edge := range horizontalEdges {
			if edge.Y == y {
				rowIntervals = append(rowIntervals, Interval{Start: edge.MinX, End: edge.MaxX})
			}
		}

		if len(rowIntervals) > 0 {
			intervals[y] = mergeIntervals(rowIntervals)
		}
	}
	timer()

	return intervals
}

func mergeIntervals(intervals []Interval) []Interval {
	if len(intervals) == 0 {
		return nil
	}

	slices.SortFunc(intervals, func(a, b Interval) int {
		return a.Start - b.Start
	})

	merged := []Interval{intervals[0]}
	for i := 1; i < len(intervals); i++ {
		last := &merged[len(merged)-1]
		curr := intervals[i]

		if curr.Start <= last.End+1 {
			if curr.End > last.End {
				last.End = curr.End
			}
		} else {
			merged = append(merged, curr)
		}
	}

	return merged
}

func getLargestAreaWithIntervals(redTiles []grid.Point, intervals map[int][]Interval) int {
	log := logger.New("Day9").With("Function", "getLargestAreaWithIntervals")

	timer := log.Timer("precompute max widths")
	maxWidthAtY := make(map[int]int)
	for y, ivs := range intervals {
		maxW := 0
		for _, iv := range ivs {
			w := iv.End - iv.Start + 1
			if w > maxW {
				maxW = w
			}
		}
		maxWidthAtY[y] = maxW
	}
	timer()

	timer = log.Timer("get potential tiles")
	type candidate struct {
		minX, maxX, minY, maxY int
		area                   int
	}

	candidates := make([]candidate, 0, len(redTiles)*(len(redTiles)-1)/2)

	for i := range redTiles {
		for j := i + 1; j < len(redTiles); j++ {
			p1, p2 := redTiles[i], redTiles[j]
			minX, maxX := p1.X, p2.X
			if minX > maxX {
				minX, maxX = maxX, minX
			}
			minY, maxY := p1.Y, p2.Y
			if minY > maxY {
				minY, maxY = maxY, minY
			}
			width := maxX - minX + 1
			height := maxY - minY + 1
			candidates = append(candidates, candidate{
				minX: minX, maxX: maxX, minY: minY, maxY: maxY,
				area: width * height,
			})
		}
	}
	timer()

	timer = log.Timer("sort potential tiles")
	slices.SortFunc(candidates, func(a, b candidate) int {
		return b.area - a.area
	})
	timer()

	timer = log.Timer("check rectangles")
	for _, c := range candidates {
		width := c.maxX - c.minX + 1

		canFit := true
		for y := c.minY; y <= c.maxY; y++ {
			if maxWidthAtY[y] < width {
				canFit = false
				break
			}
		}
		if !canFit {
			continue
		}

		if isRectangleInside(c.minX, c.maxX, c.minY, c.maxY, intervals) {
			timer()
			return c.area
		}
	}
	timer()

	return 0
}

// Gemini algorithm
// Segment Tree for Range Minimum Query
type SegmentTree struct {
	tree    []int
	yCoords []int
	yIndex  map[int]int
	size    int
}

// Gemini algorithm
func buildSegmentTree(maxWidthAtY map[int]int) *SegmentTree {
	yCoords := make([]int, 0, len(maxWidthAtY))
	for y := range maxWidthAtY {
		yCoords = append(yCoords, y)
	}
	slices.Sort(yCoords)

	yIndex := make(map[int]int, len(yCoords))
	for i, y := range yCoords {
		yIndex[y] = i
	}

	size := len(yCoords)
	tree := make([]int, 4*size)

	var build func(arrIndex, start, end int)
	build = func(arrIndex, start, end int) {
		if start == end {
			tree[arrIndex] = maxWidthAtY[yCoords[start]]
			return
		}
		mid := (start + end) / 2
		build(2*arrIndex+1, start, mid)
		build(2*arrIndex+2, mid+1, end)
		tree[arrIndex] = min(tree[2*arrIndex+1], tree[2*arrIndex+2])
	}

	if size > 0 {
		build(0, 0, size-1)
	}

	return &SegmentTree{tree: tree, yCoords: yCoords, yIndex: yIndex, size: size}
}

// Gemini algorithm
func (st *SegmentTree) Query(minY, maxY int) int {
	if st.size == 0 {
		return 0
	}

	l, okL := st.yIndex[minY]
	r, okR := st.yIndex[maxY]

	// If the exact Y is not a key, find the next available one.
	if !okL {
		// Find where minY would be inserted
		idx, _ := slices.BinarySearch(st.yCoords, minY)
		if idx >= len(st.yCoords) {
			return 0
		}
		l = idx
	}
	if !okR {
		idx, _ := slices.BinarySearch(st.yCoords, maxY)
		if idx == 0 {
			return 0
		}
		r = idx - 1
	}

	if l > r {
		// This can happen if the range is empty or falls between our yCoords.
		// A single row check might need this.
		if _, ok := st.yIndex[minY]; ok && minY == maxY {
			r = l
		} else {
			return 0
		}
	}

	var query func(arrIndex, start, end, qStart, qEnd int) int
	query = func(arrIndex, start, end, qStart, qEnd int) int {
		if qStart > end || qEnd < start {
			return math.MaxInt
		}
		if qStart <= start && qEnd >= end {
			return st.tree[arrIndex]
		}
		mid := (start + end) / 2
		leftQuery := query(2*arrIndex+1, start, mid, qStart, qEnd)
		rightQuery := query(2*arrIndex+2, mid+1, end, qStart, qEnd)
		return min(leftQuery, rightQuery)
	}

	return query(0, 0, st.size-1, l, r)
}

// Gemini algorithm
func getLargestAreaWithIntervalsOptimized(redTiles []grid.Point, intervals map[int][]Interval) int {
	log := logger.New("Day9").With("Function", "getLargestAreaWithIntervalsOptimized")

	timer := log.Timer("precompute max widths")
	maxWidthAtY := make(map[int]int)
	for y, ivs := range intervals {
		maxW := 0
		for _, iv := range ivs {
			w := iv.End - iv.Start + 1
			if w > maxW {
				maxW = w
			}
		}
		maxWidthAtY[y] = maxW
	}
	timer()

	timer = log.Timer("build segment tree")
	segTree := buildSegmentTree(maxWidthAtY)
	timer()

	timer = log.Timer("get potential tiles")
	type candidate struct {
		minX, maxX, minY, maxY int
		area                   int
	}

	candidates := make([]candidate, 0, len(redTiles)*(len(redTiles)-1)/2)

	for i := 0; i < len(redTiles); i++ {
		for j := i + 1; j < len(redTiles); j++ {
			p1, p2 := redTiles[i], redTiles[j]
			minX, maxX := p1.X, p2.X
			if minX > maxX {
				minX, maxX = maxX, minX
			}
			minY, maxY := p1.Y, p2.Y
			if minY > maxY {
				minY, maxY = maxY, minY
			}
			width := maxX - minX + 1
			height := maxY - minY + 1
			// Optimization: if a single row can't fit width, no point generating candidate
			if _, ok := maxWidthAtY[p1.Y]; ok && maxWidthAtY[p1.Y] < width {
				continue
			}
			if _, ok := maxWidthAtY[p2.Y]; ok && maxWidthAtY[p2.Y] < width {
				continue
			}

			candidates = append(candidates, candidate{
				minX: minX, maxX: maxX, minY: minY, maxY: maxY,
				area: width * height,
			})
		}
	}
	timer()

	timer = log.Timer("sort potential tiles")
	slices.SortFunc(candidates, func(a, b candidate) int {
		return b.area - a.area
	})
	timer()

	timer = log.Timer("check rectangles")
	for _, c := range candidates {
		width := c.maxX - c.minX + 1

		minWidthInRange := segTree.Query(c.minY, c.maxY)

		if minWidthInRange < width {
			continue
		}

		if isRectangleInside(c.minX, c.maxX, c.minY, c.maxY, intervals) {
			timer()
			return c.area
		}
	}
	timer()

	return 0
}

func isRectangleInside(minX, maxX, minY, maxY int, intervals map[int][]Interval) bool {
	for y := minY; y <= maxY; y++ {
		rowIntervals, exists := intervals[y]
		if !exists {
			return false
		}

		// Binary search: find the rightmost interval with Start <= minX
		lo, hi := 0, len(rowIntervals)
		for lo < hi {
			mid := (lo + hi) / 2
			if rowIntervals[mid].Start <= minX {
				lo = mid + 1
			} else {
				hi = mid
			}
		}

		// Check if the interval before 'lo' contains our range
		if lo == 0 || rowIntervals[lo-1].End < maxX {
			return false
		}
	}
	return true
}

func getLargestArea(redTiles []grid.Point) int {
	result := 0.0

	for _, point1 := range redTiles {
		for _, point2 := range redTiles {
			width := math.Abs(float64(point1.X)-float64(point2.X)) + 1
			height := math.Abs(float64(point1.Y)-float64(point2.Y)) + 1
			area := width * height
			if area > result {
				result = area
			}
		}
	}
	return int(result)
}

func getRedTiles(filename string) []grid.Point {
	file := utilities.ReadFile(filename)

	var redTiles []grid.Point
	for _, row := range file {
		coordinates := strings.Split(row, ",")
		x, _ := strconv.Atoi(coordinates[0])
		y, _ := strconv.Atoi(coordinates[1])
		redTiles = append(redTiles, grid.Point{X: x, Y: y})
	}

	return redTiles
}

File: aoc2025/rust/src/bin/day9.rs
use std::cmp::{min, max};
use std::collections::HashMap;
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;
use std::time::Instant;

#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
struct Point {
    x: i64,
    y: i64,
}

#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd)]
struct Interval {
    start: i64,
    end: i64,
}

fn main() {
    let red_tiles = get_red_tiles("files/day9.part1").unwrap();
    println!("Day 9, Part 1: {}", get_largest_area(&red_tiles));
    
    let now = Instant::now();
    let intervals = get_inside_intervals(&red_tiles);
    let part2_result = get_largest_area_with_intervals_optimized(&red_tiles, &intervals);
    let elapsed = now.elapsed();
    println!("Day 9, Part 2: {}", part2_result);
    println!("Part 2 took: {:.2?}", elapsed);
}

fn get_largest_area_with_intervals_optimized(red_tiles: &[Point], intervals: &HashMap<i64, Vec<Interval>>) -> i64 {
    let mut max_width_at_y = HashMap::new();
    for (&y, ivs) in intervals {
        let max_w = ivs.iter().map(|iv| iv.end - iv.start + 1).max().unwrap_or(0);
        max_width_at_y.insert(y, max_w);
    }

    let seg_tree = SegmentTree::new(&max_width_at_y);

    let mut candidates = Vec::new();
    for i in 0..red_tiles.len() {
        for j in i + 1..red_tiles.len() {
            let p1 = red_tiles[i];
            let p2 = red_tiles[j];
            let min_x = min(p1.x, p2.x);
            let max_x = max(p1.x, p2.x);
            let min_y = min(p1.y, p2.y);
            let max_y = max(p1.y, p2.y);
            let area = (max_x - min_x + 1) * (max_y - min_y + 1);
            candidates.push((area, min_x, max_x, min_y, max_y));
        }
    }

    candidates.sort_by_key(|k| std::cmp::Reverse(k.0));

    for (area, min_x, max_x, min_y, max_y) in candidates {
        let width = max_x - min_x + 1;
        let min_width_in_range = seg_tree.query(min_y, max_y);

        if min_width_in_range < width {
            continue;
        }

        if is_rectangle_inside(min_x, max_x, min_y, max_y, intervals) {
            return area;
        }
    }

    0
}

fn get_red_tiles(filename: &str) -> io::Result<Vec<Point>> {
    let path = Path::new(filename);
    let file = File::open(path)?;
    let reader = io::BufReader::new(file);

    let mut points = Vec::new();
    for line in reader.lines() {
        let line = line?;
        let coords: Vec<&str> = line.split(',').collect();
        if coords.len() == 2 {
            let x = coords[0].parse::<i64>().unwrap();
            let y = coords[1].parse::<i64>().unwrap();
            points.push(Point { x, y });
        }
    }
    Ok(points)
}


fn get_largest_area(red_tiles: &[Point]) -> i64 {
    let mut result = 0.0;

    for &point1 in red_tiles {
        for &point2 in red_tiles {
            let width = (point1.x - point2.x).abs() as f64 + 1.0;
            let height = (point1.y - point2.y).abs() as f64 + 1.0;
            let area = width * height;
            if area > result {
                result = area;
            }
        }
    }
    result as i64
}

fn merge_intervals(intervals: &mut Vec<Interval>) {
    if intervals.is_empty() {
        return;
    }
    intervals.sort_unstable();
    let mut merged = Vec::new();
    merged.push(intervals[0]);

    for &interval in &intervals[1..] {
        let last = merged.last_mut().unwrap();
        if interval.start <= last.end + 1 {
            last.end = max(last.end, interval.end);
        } else {
            merged.push(interval);
        }
    }
    *intervals = merged;
}

struct SegmentTree {
    tree: Vec<i64>,
    y_coords: Vec<i64>,
    size: usize,
}

impl SegmentTree {
    fn new(max_width_at_y: &HashMap<i64, i64>) -> Self {
        let mut y_coords: Vec<i64> = max_width_at_y.keys().cloned().collect();
        y_coords.sort_unstable();

        let size = y_coords.len();
        let mut tree = vec![0; 4 * size];

        if size > 0 {
            Self::build(&mut tree, &y_coords, max_width_at_y, 0, 0, size - 1);
        }

        SegmentTree { tree, y_coords, size }
    }

    fn build(tree: &mut Vec<i64>, y_coords: &[i64], max_width_at_y: &HashMap<i64, i64>, arr_index: usize, start: usize, end: usize) {
        if start == end {
            tree[arr_index] = *max_width_at_y.get(&y_coords[start]).unwrap_or(&0);
            return;
        }
        let mid = start + (end - start) / 2;
        Self::build(tree, y_coords, max_width_at_y, 2 * arr_index + 1, start, mid);
        Self::build(tree, y_coords, max_width_at_y, 2 * arr_index + 2, mid + 1, end);
        tree[arr_index] = min(tree[2 * arr_index + 1], tree[2 * arr_index + 2]);
    }

    fn query(&self, min_y: i64, max_y: i64) -> i64 {
        if self.size == 0 {
            return 0;
        }

        let l = match self.y_coords.binary_search(&min_y) {
            Ok(i) => i,
            Err(i) => i,
        };
        let r = match self.y_coords.binary_search(&max_y) {
            Ok(i) => i,
            Err(i) => if i > 0 { i - 1 } else { return 0 },
        };
        
        if l > r {
             return 0;
        }

        self.query_internal(0, 0, self.size - 1, l, r)
    }

    fn query_internal(&self, arr_index: usize, start: usize, end: usize, q_start: usize, q_end: usize) -> i64 {
        if q_start > end || q_end < start {
            return i64::MAX;
        }
        if q_start <= start && q_end >= end {
            return self.tree[arr_index];
        }
        let mid = start + (end - start) / 2;
        let left_query = self.query_internal(2 * arr_index + 1, start, mid, q_start, q_end);
        let right_query = self.query_internal(2 * arr_index + 2, mid + 1, end, q_start, q_end);
        min(left_query, right_query)
    }
}

fn get_inside_intervals(red_tiles: &[Point]) -> HashMap<i64, Vec<Interval>> {
    if red_tiles.is_empty() {
        return HashMap::new();
    }


    struct VerticalEdge { x: i64, min_y: i64, max_y: i64 }
    struct HorizontalEdge { y: i64, min_x: i64, max_x: i64 }

    let mut vertical_edges = Vec::new();
    let mut horizontal_edges = Vec::new();

    for i in 0..red_tiles.len() {
        let current = red_tiles[i];
        let next = red_tiles[(i + 1) % red_tiles.len()];

        if current.x == next.x {
            vertical_edges.push(VerticalEdge {
                x: current.x,
                min_y: min(current.y, next.y),
                max_y: max(current.y, next.y),
            });
        } else {
            horizontal_edges.push(HorizontalEdge {
                y: current.y,
                min_x: min(current.x, next.x),
                max_x: max(current.x, next.x),
            });
        }
    }

    let min_y = red_tiles.iter().map(|p| p.y).min().unwrap_or(0);
    let max_y = red_tiles.iter().map(|p| p.y).max().unwrap_or(0);

    let mut intervals = HashMap::new();
    for y in min_y..=max_y {
        let mut crossings = Vec::new();
        for edge in &vertical_edges {
            if y >= edge.min_y && y < edge.max_y {
                crossings.push(edge.x);
            }
        }
        crossings.sort_unstable();

        let mut row_intervals = Vec::new();
        for chunk in crossings.chunks_exact(2) {
            row_intervals.push(Interval { start: chunk[0], end: chunk[1] });
        }

        for edge in &horizontal_edges {
            if edge.y == y {
                row_intervals.push(Interval { start: edge.min_x, end: edge.max_x });
            }
        }

        if !row_intervals.is_empty() {
            merge_intervals(&mut row_intervals);
            intervals.insert(y, row_intervals);
        }
    }

    intervals
}

fn is_rectangle_inside(min_x: i64, max_x: i64, min_y: i64, max_y: i64, intervals: &HashMap<i64, Vec<Interval>>) -> bool {
    for y in min_y..=max_y {
        if let Some(row_intervals) = intervals.get(&y) {
            let mut lo = 0;
            let mut hi = row_intervals.len();
            while lo < hi {
                let mid = lo + (hi - lo) / 2;
                if row_intervals[mid].start <= min_x {
                    lo = mid + 1;
                } else {
                    hi = mid;
                }
            }

            if lo == 0 || row_intervals[lo - 1].end < max_x {
                return false;
            }
        } else {
            return false;
        }
    }
    true
}